src/ray/util/event.h

```diff
@@ -0,0 +1,177 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef RAY_EVENT_H_
+#define RAY_EVENT_H_
+
+#include <cmath>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+#include <boost/asio/ip/host_name.hpp>
+#include <boost/filesystem.hpp>
+#include "ray/util/logging.h"
+#include "ray/util/util.h"
+#include "src/ray/protobuf/event.pb.h"
+
+namespace ray {
+
+#define RAY_EVENT(event_type, label) \
+  ::ray::RayEvent(::ray::rpc::Event_Severity::Event_Severity_##event_type, label)
+
+class LogBasedEventReporter {
+ public:
+  LogBasedEventReporter(rpc::Event_SourceType source_type, std::string log_dir);
+  
+  virtual void Init() = 0;
+
+  virtual void Report(rpc::Event &event) = 0;
+
+  virtual void Close() = 0;
+
+  virtual std::string GetReporterKey() = 0;
+
+ protected:
+  virtual std::string EventToString(rpc::Event &event);
+ 
+ protected:
+  const std::string separator_ = "|||";
+  std::string file_name_;
+  std::string log_dir_; 
+};
+
+class EventManager {
+ public:
+  static EventManager &Instance() {
```

Comment:
There is no such shutdown or deconstrutor for EventManager.
We'd better use smart pointer for singleton instance so sources or somethings like could be released after process halted, and we can reset it in-fly.

Reply:
use clearReporters function, we can shut down the EventManager. Maybe I should modify the clearReporters to shutdown function.